pkg/tfshim/walk/walk.go

```diff
@@ -0,0 +1,198 @@
+	ret := make(SchemaPath, len(p))
+	copy(ret, p)
+	return ret
+}
+
+func (p SchemaPath) Element() SchemaPath {
+	return p.WithStep(ElementStep{})
+}
+
+func (p SchemaPath) GetAttr(name string) SchemaPath {
+	return p.WithStep(GetAttrStep{name})
+}
+
+func (p SchemaPath) WithStep(suffix SchemaPathStep) SchemaPath {
+	ret := make(SchemaPath, len(p)+1)
+	copy(ret, p)
+	ret[len(p)] = suffix
+	return ret
+}
+
+// Builds a new empty SchemaPath.
+func NewSchemaPath() SchemaPath {
+	return make(SchemaPath, 0)
+}
+
+// Finds a nested Schema at a given path.
+func LookupSchemaPath(path SchemaPath, schema shim.Schema) (shim.Schema, error) {
+	p := path
+	current := NewSchemaPath()
+	result := schema
+	for {
+		if len(p) == 0 {
+			return result, nil
+		}
+		nextResult, err := p[0].Lookup(result)
+		if err != nil {
+			return nil, fmt.Errorf("LookupSchemaPath failed at %s: %w", current, err)
+		}
+		result, p, current = nextResult, p[1:], current.WithStep(p[0])
+	}
+}
+
+// Similar to LookupSchemaPath but starts the initial step from a SchemaMap.
+func LookupSchemaMapPath(path SchemaPath, schemaMap shim.SchemaMap) (shim.Schema, error) {
+	return LookupSchemaPath(path, wrapSchemaMap(schemaMap))
+}
+
+// Represents elements of a SchemaPath.
+//
+// This interface is closed, the only the implementations given in the current package are allowed.
+type SchemaPathStep interface {
+	String() string
+	Lookup(shim.Schema) (shim.Schema, error)
+}
+
+// Drill down into an attribute by the given attribute name.
+type GetAttrStep struct {
+	Name string
+}
+
+func (step GetAttrStep) String() string {
+	return fmt.Sprintf("walk.GetAttrStep{%q}", step.Name)
+}
+
+func (step GetAttrStep) Lookup(s shim.Schema) (shim.Schema, error) {
+	if sm, ok := UnwrapSchemaMap(s); ok {
+		s, found := sm.GetOk(step.Name)
+		if !found {
+			return nil, fmt.Errorf("%s not found", step.String())
+		}
+		return s, nil
+	}
+	return nil, fmt.Errorf("%s is not applicable", step.String())
+}
+
+// Drill down into a Map, Set or List element schema.
+type ElementStep struct{}
+
+func (step ElementStep) String() string {
+	return "walk.ElementStep{}"
+}
+
+func (step ElementStep) Lookup(s shim.Schema) (shim.Schema, error) {
+	switch elem := s.Elem().(type) {
+	case shim.Resource:
+		switch s.Type() {
+		case shim.TypeMap:
+			return nil, fmt.Errorf("%s is not applicable to object types", step.String())
+		case shim.TypeList, shim.TypeSet:
+			return wrapSchemaMap(elem.Schema()), nil
+		default:
+			return nil, fmt.Errorf("%s is not applicable", step.String())
+		}
+	case shim.Schema:
+		return elem, nil
+	default:
+		return nil, fmt.Errorf("%s is not applicable", step.String())
+	}
+}
+
+func wrapSchemaMap(sm shim.SchemaMap) shim.Schema {
+	return (&schema.Schema{
+		Type: shim.TypeMap,
+		Elem: (&schema.Resource{Schema: sm}).Shim(),
+	}).Shim()
+}
+
+// Utility function to recognize nested object field type schemas encoded in shim.Schema.
+func UnwrapSchemaMap(s shim.Schema) (shim.SchemaMap, bool) {
```

Comment:
Does this need to be public?

Reply:
Eh, probably not from this package and the naming is off. I will hide.